File: LeetCode/p20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        match = {'(': ')', '[': ']', '{': '}'}
        for c in s:
            if c in ['(', '[', '{']:
                stack.append(c)
            elif not stack or match[stack.pop()] != c:
                return False
        return not stack



if __name__ == '__main__':
    # begin
    s = Solution()
    print(s.isValid("([)]{}"))

File: LeetCode-Java/p20.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> match = new HashMap<>();
        match.put('(', ')');
        match.put('[', ']');
        match.put('{', '}');

        for (char c : s.toCharArray()) {
            // If it's an opening bracket, push onto the stack
            if (match.containsKey(c)) {
                stack.push(c);
            }
            // If it's a closing bracket, check for a matching opening bracket
            else if (stack.isEmpty() || match.get(stack.pop()) != c) {
                return false;
            }
        }
        // If the stack is empty, all brackets were matched correctly
        return stack.isEmpty();
    }
}